crates/runc/src/lib.rs

```diff
@@ -346,31 +340,54 @@ impl Runc {
     }
 }
 
+// a macro tool to cleanup the file with name $filename,
+// there is no async drop in async rust, so we have to call remove_file everytime
+// after a temp file created, before return of a function.
+// with this macro we don't have to write the match case codes everytime.
+#[cfg(feature = "async")]
+macro_rules! tc {
+    ($b:expr, $filename: expr) => {
+        match $b {
+            Ok(r) => r,
+            Err(e) => {
+                let _ = tokio::fs::remove_file($filename).await;
+                return Err(e);
+            }
+        }
+    };
+}
+
+#[cfg(not(feature = "async"))]
+pub trait Executor {
+    fn execute(&self, cmd: Command) -> Result<(ExitStatus, u32, String, String)>;
+}
+
+#[cfg(feature = "async")]
+#[async_trait]
+pub trait Executor {
+    async fn execute(&self, cmd: Command) -> Result<(ExitStatus, u32, String, String)>;
+}
+
```

Comment:
I would prefer to have two `execute` implementations depending on `async` feature toggle. There is no need to have `Executor` trait here (moreover `async_trait` adds additional overhead to this).

Reply:
Actually the reason I add Executor trait here is not because the Command is different from sync to async. It is actually because the `wait` function of `Command` may block because runc-shim may start a child reaper, which will catch the SIGCHLD signal, that makes `wait` function can not receive the SIGCHLD, and wait there forever. As the runc crate can not depends on the shim crate, I can only expose a `Executor` trait in runc crate, and let runc-shim to implements their own `execute` logic based on the monitor from shim. the details is in implementation for the `ShimExecutor`. 
And I think the Monitor should be a helper struct for all shim implementation, so I think it's better if it is in the shim crate.

Reply:
Same problem occurs in Sync version too. Most of [integration test](https://github.com/containerd/rust-extensions/pull/37#issuecomment-1059849843) failed with `Others("Other: unknown error after kill: error occured in runc: no child processes (os error 10)"): unknown` may caused by runc-shim process have set child subreaper. I also have found that everything's ok on CentOS but wrong with Ubuntu. So I think we still need to keep this `Executor` in Sync version.

Do we have a better solution?

Reply:
I see what you're trying to do.
You want to inject custom logic before/after spawning the process.
It might cleaner to allow extensions via runc's Builder.

Something to inspire from Tokio:
https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/builder.rs#L68
https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/builder.rs#L351

```rust
runc::builder().custom_spawner(...) ?
```

WDT?